pkg/datapath/linux/devices_controller.go

```diff
@@ -0,0 +1,653 @@
+	// addr/route/link updates before processing the batch.
+	batchingDuration = 100 * time.Millisecond
+
+	// restartWaitDuration is the amount of time to wait after
+	// a netlink failure before restarting from scratch.
+	restartWaitDuration = time.Second
+
+	excludedDevicePrefixes = []string{
+		defaults.HostDevice,
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+		"sit",
+	}
+
+	// Route filter to look at all routing tables.
+	routeFilter = netlink.Route{
+		Table: unix.RT_TABLE_UNSPEC,
+	}
+	routeFilterMask = netlink.RT_FILTER_TABLE
+)
+
+type DevicesConfig struct {
+	// Devices is the user-specified devices to use. This can be
+	// either concrete devices ("eth0,eth1"), or a wildcard "eth+".
+	// If empty the devices are auto-detected according to rules defined
+	// by isSelectedDevice().
+	Devices []string
+}
+
+type devicesControllerParams struct {
+	cell.In
+
+	Config      DevicesConfig
+	Log         logrus.FieldLogger
+	DB          statedb.DB
+	DeviceTable statedb.Table[*tables.Device]
+	RouteTable  statedb.Table[*tables.Route]
+
+	// netlinkFuncs is optional and used by tests to verify error handling behavior.
+	NetlinkFuncs *netlinkFuncs `optional:"true"`
+}
+
+type devicesController struct {
+	params devicesControllerParams
+	log    logrus.FieldLogger
+
+	initialized    chan struct{}
+	filter         deviceFilter
+	l3DevSupported bool
+
+	cancel context.CancelFunc // controller's context is cancelled when stopped.
+}
+
+func newDevicesController(lc hive.Lifecycle, p devicesControllerParams) *devicesController {
+	dc := &devicesController{
+		params:      p,
+		initialized: make(chan struct{}),
+		filter:      deviceFilter(p.Config.Devices),
+		log:         p.Log,
+	}
+	lc.Append(dc)
+	return dc
+}
+
+func (dc *devicesController) Start(startCtx hive.HookContext) error {
+	if dc.params.NetlinkFuncs == nil {
+		var err error
+		dc.params.NetlinkFuncs, err = makeNetlinkFuncs(netns.None())
+		if err != nil {
+			return err
+		}
+
+		// Only probe for L3 device support when netlink isn't mocked by tests.
+		dc.l3DevSupported = probes.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbChangeHead) == nil
+	}
+
+	var ctx context.Context
+	ctx, dc.cancel = context.WithCancel(context.Background())
+
+	go dc.run(ctx)
+
+	// Wait until the initial population of the tables has finished
+	// successfully or the start has been aborted.
+	select {
+	case <-dc.initialized:
+	case <-startCtx.Done():
+		dc.cancel()
+	}
+
+	return nil
+}
+
+func (dc *devicesController) run(ctx context.Context) {
+	defer dc.params.NetlinkFuncs.Close()
+
+	// Run the controller in a loop and restarting on failures until stopped.
+	// We're doing this as netlink is an unreliable protocol that may drop
+	// messages if the socket buffer is filled (recvmsg returns ENOBUFS).
+	for ctx.Err() == nil {
+		dc.subscribeAndProcess(ctx)
+
+		t, stop := inctimer.New()
+
+		select {
+		case <-ctx.Done():
+			stop()
+			return
+		case <-t.After(restartWaitDuration):
```

Comment:
I'm guessing there's no need for a backoff here?

Reply:
I removed the backoff in the end as it doesn't really bring much value, plus the logic for resetting the backoff would've been a bit involved (when would we even reset it?).